### PR TITLE
Feat/user roles permissions gems

### PR DESCRIPTION
**Pull request by stefanjbloom**
## Type of Change
- [x] feature ⛲
- [x] documentation update 📃
- [x] refactor 🧑‍💻
- [x] testing 🧪

## Description
This branch incorporates a roles system and authorization based off of what those roles are.  It combines the JWT and Rolify gem for authentication and users role-assignment as well as Pundit gem for authorization based off of a users role.  

## Motivation and Context
We needed to add authentication, user roles and authorizations to the app.  This branch will involve many devs here having to refactor their controllers, request spec tests and create new policies for their models controllers.

## Related Tickets
closes #12 

Any ticket for a controller action will need refactoring. See README

## Added Test?
- [x] Yes 🫡
Unit, Model, Controller, Policy Tests

- [ ] All previous tests still pass 🥳
    - companies_controller spec tests are now all failing.  Refactored users_controller spec tests as a template for all future refactors.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: app/models/role.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class Role < ApplicationRecord
+  has_and_belongs_to_many :users, :join_table => :users_roles
+  
+  belongs_to :resource, polymorphic: true, optional: true
+
```
Comment (stefanjbloom) [inline app/models/role.rb:5]: File generated by Rolify.  Belongs to resource indicates the optional scalability of assigning polymorphic associations to a user depending if we want them to have controlled access to editing resources. (like creating a moderator role)

File: app/policies/company_policy.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  # code, beware of possible changes to the ancestors:
+  # https://gist.github.com/Burgestrand/4b4bc22f31c8a95c425fc0e30d7ef1f5
+
+  class Scope < ApplicationPolicy::Scope
+    # NOTE: Be explicit about which records you allow access to!
+    # def resolve
+    #   scope.all
+    # end
+  end
+end
```
Comment (stefanjbloom) [inline app/policies/company_policy.rb:14]: Placeholder for policy regarding companies_controller

File: app/policies/contact_policy.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  # code, beware of possible changes to the ancestors:
+  # https://gist.github.com/Burgestrand/4b4bc22f31c8a95c425fc0e30d7ef1f5
+
+  class Scope < ApplicationPolicy::Scope
+    # NOTE: Be explicit about which records you allow access to!
+    # def resolve
+    #   scope.all
+    # end
+  end
+end
```
Comment (stefanjbloom) [inline app/policies/contact_policy.rb:14]: Placeholder for policy regarding contacts_controller

File: app/policies/job_policy.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  # code, beware of possible changes to the ancestors:
+  # https://gist.github.com/Burgestrand/4b4bc22f31c8a95c425fc0e30d7ef1f5
+
+  class Scope < ApplicationPolicy::Scope
+    # NOTE: Be explicit about which records you allow access to!
+    # def resolve
+    #   scope.all
+    # end
+  end
+end
```
Comment (stefanjbloom) [inline app/policies/job_policy.rb:14]: Placeholder for policy regarding jobs_controller

File: config/initializers/rolify.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Rolify.configure do |config|
+  # By default ORM adapter is ActiveRecord. uncomment to use mongoid
+  # config.use_mongoid
+
+  # Dynamic shortcuts for User class (user.is_admin? like methods). Default is: false
+  # config.use_dynamic_shortcuts
+
```
Comment (stefanjbloom) [inline config/initializers/rolify.rb:7]: Explicitly NOT using dynamic shortcuts/built-in provided by rolify since they allow a User to have more than one role.  A user should ONLY have one role.

**Comment by stefanjbloom** [comment]
Does anyone NOT understand what's going on with the inclusion of these gems and the implications for this project?  I am absolutely available to answer how this effects this project as a whole and am open to leading a meeting regarding this new and critical functionality.

File: README.md
Comment (Sgalvin36) [inline README.md:1]: Very helpful explanation for everything. Something that may be helpful is to create a ticket on the backlog for each of the controllers that this impacts and/or the parts that will need reworking, so as to not leave it up to the people that were working on those parts to remember to go back and readjust.

File: app/controllers/api/v1/sessions_controller.rb [hunk 6]
```diff
@@ -1,10 +1,12 @@
 module Api
   module V1
     class SessionsController < ApplicationController
+      before_action :skip_authorization
```
Comment (Sgalvin36) [inline app/controllers/api/v1/sessions_controller.rb:4]: Is skip_authorization the same as just not having anything in the before_action?
If the skip_authorization wasn't added, wouldn't we be able to access this controller at any time, which is what we want since we're using this to help our authentication process?

File: app/controllers/api/v1/users_controller.rb
Comment (Sgalvin36) [inline app/controllers/api/v1/users_controller.rb:1]: Are we needing to declare the `authorize` for every action? Shouldn't we need to know who the user is before we authorize? On lines 5 and 15 it looks like we're authorizing the User table instead of a specific user is why I ask.

File: app/models/user.rb
Comment (Sgalvin36) [inline app/models/user.rb:1]: I appreciate the little catch when assigning new roles to destroy old roles first. Certainly had my fair share of experiences with applications where once you got assigned a role it was yours forever, even if it wasn't a role you needed anymore.

File: config/initializers/rolify.rb (hunk 5 above)
Comment (Sgalvin36) [reply config/initializers/rolify.rb:7]: I would say early stages yes, however I can think of times where as applications grow, individual users can have many roles (i.e. think of moderators in chats, they're still chatters themselves but also have the privileges of what moderating gives them). So certainly something to keep in mind as this application grows

**Review by Sgalvin36 — APPROVED** [review]
This is looking really good. It was a big task to handle incorporating authentication into an application. That was no easy feat, so thank you for the work that you've done with this!!

File: config/initializers/rolify.rb (hunk 5 above)
Comment (stefanjbloom) [reply config/initializers/rolify.rb:7]: Good point, and I agree.  A user can potentially have a role as a :user for everywhere except in companies, where they may be a :moderator.  As it stands now we just have :user (default state) and :admin.  Absolutely can grow to have more roles.

File: app/controllers/api/v1/sessions_controller.rb (hunk 6 above)
Comment (stefanjbloom) [reply app/controllers/api/v1/sessions_controller.rb:4]: We include Pundit via **ApplicationController** `include Pundit::Authorization`, which is inherited in every child controller.  Pundit automatically checks for authorization, based on users role, for every single controller action.

Since **SessionsController**'s SRP is creating a session and generating a token, it should not overlap with authorization at all, just authentication.  So it makes sense to put Pundit's `skip_authorization` method in **SessionsController** so that Pundit won't verify authorization in a place it shouldn't/does not exist yet.  

Simply put, authorization checks are irrelevant in this controller b/c the user is technically not even authenticated yet. Pundit will throw an error if `before_action :skip_authorization` isn't included in this particular controller.
`Pundit::AuthorizationNotPerformedError:
       Api::V1::CompaniesController`

To extrapolate further, we should not even create a policy for **SessionsController**.

Does that answer the question?

File: README.md
Comment (stefanjbloom) [reply README.md:1]: 2 new issues for created to address this, thank you!
